### PR TITLE
GLTFExporter: Serialize userData.gltfExtensions

### PR DESCRIPTION
**Pull request by robertlong**
The GLTFLoader loads unknown extensions onto the `userData.gltfExtensions` property. This PR adds support for exporting those unknown extensions or any other extensions on the `userData.gltfExtensions` property.

We plan on using this feature in the Mozilla Hubs editor in order to support serializing component data and round-tripping GLTF models through the GLTFLoader and GLTFExporter.

### CODE REVIEW

**Review by donmccurdy — APPROVED** [review]
Looks good to me, /cc @takahirox

**Comment by takahirox** [comment]
A concern. If unknown extensions have indices referring to glTF nodes/textures/materials and etc.,  no one knows if they point the proper entities in the exported glTF.

**Comment by robertlong** [comment]
I think that’s a fair point, but how can we validate indices in unknown extensions? I think this feature is simply a way for developers to implement their own vendor extension implementations or at least to make imported glTF scenes able to be exported again. Validation of indices will have to be handled by the developers implementing these features.

**Comment by takahirox** [comment]
One of the most important things for glTF exporter is to export valid glTF (we've discussed before). Viewers could fail to load glTF with wrong extension parameter. So I don't prefer this feature as default.

Adding dev option to the exporter to enable this feature would be ok, I guess?

**Comment by donmccurdy** [comment]
Sure, something like an `includeExtensions` parameter would be OK with me. 👍

**Comment by robertlong** [comment]
Sounds good. I've added the `includeCustomExtensions` option.

**Comment by fernandojsg** [comment]
Sorry for the late reply. But yeah I agree with the optional parameter for that :+1:

**Review by fernandojsg — APPROVED** [review]

File: docs/examples/exporters/GLTFExporter.html [hunk 1]
```diff
@@ -84,6 +84,7 @@ <h3>[method:null parse]( [param:Object3D input], [param:Function onCompleted], [
 			<li>animations - Array<[page:AnimationClip AnimationClip]>. List of animations to be included in the export.</li>
 			<li>forceIndices - bool. Generate indices for non-index geometry and export with them. Default is false.</li>
 			<li>forcePowerOfTwoTextures - bool. Export with images resized to POT size. This option works only if embedImages is true. Default is false.</li>
+			<li>includeCustomExtensions - bool. Export custom glTF extensions defined on an object's <em>userData.gltfExtensions<em> property. Default is false.</li>
```
Comment (takahirox) [inline docs/examples/exporters/GLTFExporter.html]: You're missing `/` in closing `</em>`.

File: examples/js/exporters/GLTFExporter.js [hunk 2]
```diff
@@ -331,21 +332,50 @@ THREE.GLTFExporter.prototype = {
...
+					for ( var extensionName in json.gltfExtensions ) {
+
+						gltfProperty.extensions[ extensionName ] = json.gltfExtensions[ extensionName ];
+						extensionsUsed[ extensionName ] = true;
+
+					}
+
+				}
+
+				delete json.gltfExtensions;
```
Comment (takahirox) [inline examples/js/exporters/GLTFExporter.js]: I'd like to make clear the behavior where `.includeCustomExtensions` is `false` and `userData` has `.gltfExtensions`.

1. Export `userData.gltfExtensions` as `extra.gltfExtensions`.
2. Don't export `userData.gltfExtensions`.

I prefer 1.

File: examples/js/exporters/GLTFExporter.js [hunk 3]
```diff
@@ -1684,6 +1704,8 @@ THREE.GLTFExporter.prototype = {
 
 			}
 
+			serializeUserData( scene, gltfScene );
```
Comment (takahirox) [inline examples/js/exporters/GLTFExporter.js]: Oops, we haven't exported `scene.userData`. Good fix.

**Comment by robertlong** [comment]
@takahirox I finally had some time to look at this PR. Sorry for the long wait. I think the small change I made should satisfy your concerns.

File: examples/js/exporters/GLTFExporter.js [hunk 4]
```diff
@@ -331,21 +332,50 @@ THREE.GLTFExporter.prototype = {
 		 * Serializes a userData.
 		 *
 		 * @param {THREE.Object3D|THREE.Material} object
-		 * @returns {Object}
+		 * @param {Object} gltfProperty
 		 */
-		function serializeUserData( object ) {
+		function serializeUserData( object, gltfProperty ) {
+
+			if ( ! object.userData || Object.keys( object.userData ).length === 0 ) {
```
Comment (mrdoob) [inline examples/js/exporters/GLTFExporter.js]: Minor thing. Why are we checking `object.userData`? Shouldn't `object.userData` always be there?

https://github.com/mrdoob/three.js/blob/dev/src/core/Object3D.js#L99

Comment (donmccurdy) [reply examples/js/exporters/GLTFExporter.js]: The object may also be a `geometry`, and I think A-Frame needed to use a new version of GLTFExporter with an older version of three.js for a while. The most recent A-Frame release is on r101, so perhaps it isn't necessary now.

**Review by donmccurdy — APPROVED** [review]

File: examples/js/exporters/GLTFExporter.js (hunk 4 above)
Comment (robertlong) [reply examples/js/exporters/GLTFExporter.js]: So is the check for `object.userData` still necessary or can I remove it?

Comment (donmccurdy) [reply examples/js/exporters/GLTFExporter.js]: It can be removed. 👍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**